### PR TITLE
cl12: Backport spirv_new from cl21_trunk to test cl_khr_il_program

### PR DESCRIPTION
**Pull request by aarongreig**
This PR backports the spirv_new tests from the 2.1 cts to test the `cl_khr_il_program` extension. The new tests have been added to the test `.csv`s under the "Extensions" heading.
Pending spirv_new PRs not included in this one (such as #119) will be backported separately.

### CODE REVIEW

File: test_conformance/spirv_new/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,69 @@
+######################################################################################################
```
Comment (aarongreig) [inline test_conformance/spirv_new/CMakeLists.txt]: this cmake is not a backport, it was written pretty much from scratch due to incompatibility between cl21 cmake and cl12 cmake

File: test_common/harness/kernelHelpers.h [hunk 2]
```diff
@@ -121,6 +126,21 @@ cl_device_fp_config get_default_rounding_mode( cl_device_id device );
...
     }
 
+#define PASSIVE_REQUIRE_FP16_SUPPORT(device)                            \
+    if (!is_extension_available(device, "cl_khr_fp16"))                 \
+    {                                                                   \
+        log_info("\n\tNote: device does not support fp16. Skipping test...\n"); \
+        return 0;                                                       \
+    }
+
+#define PASSIVE_REQUIRE_IL_PROGRAM_SUPPORT( device )    \
```
Comment (aarongreig) [inline test_common/harness/kernelHelpers.h]: this is new to support SPIR-V being an extension where no such check is necessary in 2.1

File: test_conformance/spirv_new/main.cpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+    std::vector<unsigned char> result(tmpBuffer.begin(), tmpBuffer.end());
+
+    return result;
+}
+
+
+std::vector<unsigned char> readSPIRV(const char* file_name, const cl_uint address_bits)
+{
+    const char* addrWidth = address_bits == 32 ? "32" : "64";
```
Comment (aarongreig) [inline test_conformance/spirv_new/main.cpp]: this file has some additions (starting with this line) from #172

File: test_conformance/spirv_new/procs.h [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        }                                               \
+    };                                                  \
+    test_##name##_class *var_##name =                   \
+        createAndRegister<test_##name##_class>(#name);  \
+    int test_##name(cl_device_id deviceID,              \
+                    cl_context context,                 \
+                    cl_command_queue queue,             \
+                    int num_elements)
+
+std::vector<unsigned char> readSPIRV(const char *file_name, const cl_uint address_bits);
```
Comment (aarongreig) [inline test_conformance/spirv_new/procs.h]: redefined in #172

File: test_conformance/spirv_new/procs.h [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    virtual basefn getFunction() = 0;
+};
+
+// parseAndCallCommandLineTests called by runTestHarness expects there to
+// be an additinal test at the end of the get list called "all", if the
+// user requests "all" on the command line it will not call the specific
+// test function but instead run all the tests. This is required to avoid
+// off by 1 failures when the user requested the last test in the test list
+// by name on the command line.
+class test_all_class : public baseTestClass
```
Comment (aarongreig) [inline test_conformance/spirv_new/procs.h]: this and the similarly commented change in `spirv_new/main.cpp` are additions to fix the issue described above

File: test_conformance/spirv_new/types.hpp [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+Tv notOpVec(Tv in)
+{
+    Tv out;
+    for (int i = 0; i < N; i++) {
+        out.s[i] = ~in.s[i];
+    }
+    return out;
+}
+
+template <typename T>
```
Comment (aarongreig) [inline test_conformance/spirv_new/types.hpp]: this definition and changes to use of `vector` in a lot of the test files are from #177

**Comment by aarongreig** [comment]
just pushed the updated commit (third times the charm)

**Comment by aarongreig** [comment]
is there still interest in moving forward with these changes as a means to reduce the diff between cl12 and the other branches?

**Comment by kpet** [comment]
Yes but I'm tempted to suggest we wait until the `cl2x_trunk` are merged into `master` to not increase further the number of branches we need to port individual patches to.

**Comment by kpet** [comment]
@aarongreig Thinking about this a bit more, I think we'd better spend the time on enabling 1.2 on the `master` branch rather than porting `spirv_new` to `cl12_trunk`. This is the long-term direction anyway. What do you think?

**Comment by kpet** [comment]
As commented on #479, we decided to make master suitable for 1.2 implementations instead and we need to be able to test the core feature and extension from the same binary, all working towards #259. Closing this PR.

**Pull request closed without merging**